packages/kbn-observable/README.md

```diff
@@ -0,0 +1,48 @@
+# k$
+
+Experiment to see how we might get some of the benefits of a library like RxJS
+using nothing but the native `Observable` object which is proposed by
+https://github.com/tc39/proposal-observable.
+
+## Example:
+
+```js
+import { $from, k$, map, last } from 'kbn-observable';
+
+// Calling $from will attempt to cast the input value
+// to an observable:
+// 
+// - native observable: returns input value
+// - non-native observable: wraps in a native observable
+// - promise: observable that produces a single value or error
+// - iterable (except strings): emits all items from iterable
+// 
+const observable = $from(1, 2, 3);
+
+// `k$` accepts a second argument, an array of "operators" that modifies the
+// input value and return an observable that reflects all of the modifications.
+k$(observable, [ map(i => 2017 + i), last() ])
```

Comment:
I very much prefer the two-argument signature to the n-argument signature you've implemented here, mostly so the observable and operators are clearly separated but also so that multi line calls look a *little* more like chaining:

```js
const object$ = $fromFactory(async() => {
  const resp = await ajax();
  return resp.objects;
});

k$(object$, [
  map(...),
  filter(...),
  reduce(...)
]);

// vs

k$(
  object$,
  map(...),
  filter(...),
  reduce(...)
);
```

Reply:
I agree. I did that initially, but TypeScript struggles with typing it for some reason, e.g. if you map from a string to a number in the first operator, the second operator will still be typed as going from string (not number). Might have been something weird I did, so I can try to revert that and see if I can make it work.

Reply:
> TypeScript struggles with typing it for some reason
😭

Reply:
Yeah, I know :/

I played around with this, and it doesn't seem to work with the array. [Simplified example using TypeScript playground](https://www.typescriptlang.org/play/#src=type%20UnaryFn%3CT%2C%20R%3E%20%3D%20(x%3A%20T)%20%3D%3E%20R%3B%0D%0A%0D%0A%2F%2F%20Overloads%0D%0Aexport%20function%20createPipe%3CA%2C%20B%3E(ops%3A%20%5BUnaryFn%3CA%2C%20B%3E%5D)%3A%20UnaryFn%3CA%2C%20B%3E%3B%0D%0Aexport%20function%20createPipe%3CA%2C%20B%2C%20C%3E(ops%3A%20%5BUnaryFn%3CA%2C%20B%3E%2C%20UnaryFn%3CB%2C%20C%3E%5D)%3A%20UnaryFn%3CA%2C%20C%3E%3B%0D%0Aexport%20function%20createPipe%3CA%2C%20B%2C%20C%2C%20D%3E(ops%3A%20%5BUnaryFn%3CA%2C%20B%3E%2C%20UnaryFn%3CB%2C%20C%3E%2C%20UnaryFn%3CC%2C%20D%3E%5D)%3A%20UnaryFn%3CA%2C%20D%3E%3B%0D%0A%0D%0Aexport%20function%20createPipe%3CT%2C%20R%3E(%0D%0A%20%20ops%3A%20Array%3CUnaryFn%3CT%2C%20R%3E%3E%0D%0A)%3A%20UnaryFn%3CT%2C%20R%3E%20%7B%0D%0A%20%20if%20(ops.length%20%3D%3D%3D%201)%20%7B%0D%0A%20%20%20%20return%20ops%5B0%5D%3B%0D%0A%20%20%7D%0D%0A%0D%0A%20%20return%20(input%3A%20T)%3A%20R%20%3D%3E%20ops.reduce((prev%3A%20any%2C%20fn)%20%3D%3E%20fn(prev)%2C%20input)%3B%0D%0A%7D%0D%0A%0D%0Aconst%20toNumber%20%3D%20(str%3A%20string)%20%3D%3E%20parseInt(str%2C%2010)%3B%0D%0Aconst%20toString%20%3D%20(num%3A%20number)%20%3D%3E%20num.toString()%3B%0D%0A%0D%0Aconst%20process%20%3D%20createPipe(%5BtoNumber%2C%20toString%2C%20toNumber%5D)%3B%0D%0A%0D%0Aconst%20result%3A%20number%20%3D%20process(%22123%22)%3B%0D%0A). To display the error I'm just trying to "force" `result` to be a number, but you'll see that TypeScript believes it should be a string instead. It's probably related to how TypeScript handles overloads and generics within arrays.

Reply:
It can also be something like this or similar :)
```ts
k$(object$)(
  map(...),
  filter(...),
  reduce(...)
);
```

Reply:
Hmm, interesting idea @azasypkin

Reply:
Yeah, it's an interesting idea. I wonder if people will just start doing something like:

```js
const myFactory$ = k$(object$);
```

And then we're in some ways kinda back where we started. However, this is a very simple type, and the result of calling `myFactory$` above will always be a native observable.

I'll go through and make the changes, then we can see if we like it.